Completed error logging to Audit (D990)

droberts-ctrlo opened the pull request.

droberts-ctrlo committed:
Completed error logging to Audit

Also add logging upload to logging JS

Oliver-ctrlo commented on lib/GADS/Audit.pm:
```diff
@@ -70,7 +70,26 @@ has filtering => (
 );
 
-sub audit_types{ [qw/user_action login_change login_success logout login_failure/] };
+sub audit_types{ [qw/user_action login_change login_success logout login_failure script_error/] };
```
Can we have new lines for each here?

Oliver-ctrlo commented on src/frontend/js/lib/logging.js:
```diff
@@ -1,34 +1,62 @@
+        // This is wrapped so that anything that's not an object is converted to a string
+        output += `${message[i]}`;
       }
+      if (i < message.length - 1) output += ' ';
```
Why do we need the extra space here at the end of the message?

Oliver-ctrlo replied:
Consider using `message.join`.

Oliver-ctrlo commented on src/frontend/js/lib/util/scriptErrorHandler/index.ts:
```diff
@@ -0,0 +1,35 @@
+import { logging } from "logging";
+import { uploadMessage } from "./lib/MessageUploader";
+
+const createErrorString = (message: string, source: any, lineno: number, colno: number, error: Error | string | null) => {
```
Can we have new lines here as-well?

droberts-ctrlo replied:
No - convention in JS is single-line for methods (this is a constant method).

droberts-ctrlo committed:
Update to use `Log::Report`

abeverley commented on lib/GADS/API.pm:
```diff
@@ -1464,6 +1464,15 @@ get '/api/get_key' => require_login sub {
+post '/api/script_error' => require_login sub {
+    my $body = _decode_json_body();
+
+    info "SCRIPT ERROR: " . $body->{url} . " - " . $body->{description};
```
Might as well use the available interpolation:
```
info __x"SCRIPT ERROR: {url} - {description}",
    url => $body->{url}, description => $body->{description};
```

abeverley commented on lib/GADS/API.pm:
```diff
@@ -1464,6 +1464,15 @@ get '/api/get_key' => require_login sub {
+    info "SCRIPT ERROR: " . $body->{url} . " - " . $body->{description};
+
+    content_type 'application/json; charset=UTF-8';
+    return success "Script error logged successfully";
```
If you haven't already, can you test that this does what you think - I am not sure what will be returned from the endpoint with this code.

droberts-ctrlo replied:
I don't quite know why I put `success` there - I've removed it now.

droberts-ctrlo committed:
Change code as per review

Also remove extrenous logging and debugging code that was added during development of this feature.

droberts-ctrlo committed:
Fix erroring test

abeverley commented on lib/GADS/API.pm:
```diff
@@ -1464,6 +1464,16 @@ get '/api/get_key' => require_login sub {
+        url => $body->{url}, description => $body->{description};
+
+    content_type 'application/json; charset=UTF-8';
+    return "Script error logged successfully";
```
Sorry, one final thing - you are setting the content-type to be JSON but not returning JSON. It doesn't matter which way around you fix it, you just need to ensure the content-type header and the actual content type match.

droberts-ctrlo replied:
And *that's* why I used `success` - it was a typo, it was supposed to be `_success`!

droberts-ctrlo committed:
Return JSON as required
